rfcs/2022-11-04-automatic-component-validation.md

```diff
@@ -0,0 +1,501 @@
+  actually lowercased all strings, etc.
+- Validating any component types besides sources, transforms, and sinks.
+
+## Pain
+
+Currently, there is a significant amount of pain around knowing that a component emits the
+intended/relevant telemetry for that component, and that the values emitted are correct. As
+developers providing support, and operators actually running Vector, the ability to build dashboards
+and alerting that work consistently for different component types is of the utmost importance.
+
+The current process of validating components is inherently manually: we can write a unit test to
+check for compliance, but if we forget to write that unit test, or don't update all unit tests when
+we want to change the behavior for, say, all sources... then our compliance checking can quickly
+become out-of-sync with reality.
+
+Humans are also inherently fallible: it is easy to look at data and misread a number, to think
+something is there when it isn't. This has occurred multiple times -- as evidenced by the multiple
+issues linked in the **Context** section -- during work to validate components, despite multiple
+engineers participating in the work, each of them giving it a fresh look.
+
+## Proposal
+
+### User Experience
+
+Instead of having to manually add unit tests to an individual component, components would be
+automatically tested if they were usable. Being "usable", in this case, means that the component is
+present in a given Vector binary: if we're building Vector in a way that includes, say, component X,
+its mere inclusion in the binary will ensure that component X gets tested. It should effectively be
+impossible to include a component in a build in a way that doesn't surface it to be tested.
+
+These components would be validated for compliance by being run automatically in a tailored test
+harness called the "runner." This runner will be able to take any eligible component -- source,
+transform, or sink -- and build an instance of it that we would then run. The runner would control
+the input events, as well as any external resource -- HTTP server that gets scraped, downstream TCP
+socket that gets sent to, etc -- such that we would know exactly what events were sent as input and
+what events were received as output. Additionally, based on the structure of the runner, we would be
+able to isolate and capture any events/metrics emitted by the component, and have a full accounting
+of the related internal telemetry at the end of a validation run.
+
+This runner would itself be invoked as a normal unit test in Vector, doing a validation run for any
+component currently contained in the test binary. A command would be added to allow developers to
+run these validation tests locally, but a CI check would also be added to ensure the validation
+tests were run on a binary that resembles an actual release, such that we would be ensuring the
+compliance of all components that can be used.
```

Comment:
It will always run in CI, whether for a PR or an official versioned release. The point is more that: the test mechanism can only test components built into the test binary... so we need to build that binary like an official versioned release: all feature flags enabled, etc.

Reply:
Ah, gotcha. :thought_balloon:  I didn't take away that from the initial reading of the sentence, but I'm also struggling to find a better way to phrase it :laughing:  and, maybe it's just a lack of familiarity on my part.